### ISSUE TITLE
Lagrange freezes when opening an new URL while it's loading

### ISSUE THREAD

**Issue by rapenne-s** [skyjake/lagrange#302]
Hi,

I'm using Lagrange 1.5.2 on OpenBSD amd64, I found  a bug I can trigger reliably.

If you open a new URL in the current tab while Lagrange is loading, it will freeze. This is reproducible by clicking on bookmarks twice or different bookmarks quickly. It also works with "subscribed bookmarks" panel or even using the URL input (you need some dexterity skill to trigger using this way).

Lagrange requires a kill -9 to completely stops otherwise the window can be destroyed with xkill but a lagrange process will stay opened without a GUI associated.

**Comment by skyjake** [comment]
Sounds like one of the network request threads get stuck. I'll need to try to reproduce this in a VM; at least macOS does not have this issue.

**Comment by rapenne-s** [comment]
I've been able to reproduce the issue on Linux but it unstuck itself after some time between 5 or 10 seconds (approximately), it doesn't use unusual CPU while stuck.

I tried again on OpenBSD but I waited, however it stays stuck forever.

**Comment by skyjake** [comment]
[This commit](https://git.skyjake.fi/skyjake/the_Foundation/commit/f0f2d481e65dd8bdfcd4141486b56accbd713f2f) may or may not affect this issue.

**Comment by skyjake** [comment]
Status update: I haven't been able to reproduce this on any OS, but I haven't tried on OpenBSD yet.

**Comment by rapenne-s** [comment]
I tried to import the commit on my 1.5.2 build and it didn't change the behavior.

I can reproduce the issue on Linux but it's harder, I had to click a lot between different bookmarks to get it broken.
As OpenBSD is slower in syscall execution, there may be a race condition somewhere that is much more noticeable on OpenBSD because it's based on time, so on Linux except if you really click too much you don't notice it while on OpenBSD I trigger it very often when changing from bookmark to bookmark.

**Comment by skyjake** [comment]
> there may be a race condition somewhere

I tried this in OpenBSD 6.9 in a virtual machine, but the overall X11 performance was so bad that race conditions might've been resolved by that. Could not reproduce any hangs.

**Comment by skyjake** [comment]
Please test [https://github.com/skyjake/lagrange/releases/tag/v1.6.3](v1.6.3) and see if this freezing still occurs.

**Comment by rapenne-s** [comment]
I can still reproduce the bug on 1.6.3 on OpenBSD (development version, future 7.0) amd64

**Comment by skyjake** [comment]
Thanks. I'll keep trying to repro...

**Comment by skyjake** [comment]
I am now able to reproduce this bug. VMs weren’t helpful, but I installed OpenBSD 6.9 on an external drive and am now running it on real hardware.
 
Let’s see if I can figure out what’s going wrong here…

**Comment by skyjake** [comment]
I believe I've found and fixed this bug (still running OpenBSD 6.9). The app would freeze because it was waiting for a thread to exit that had already been stopped and joined before. The fix is in the dev branch (well, in the updated the_Foundation submodule).

Thanks to this foray into OpenBSD, I also discovered and fixed two other issues:

* Crash when clicking on sidebar items due to accessing invalidated pointers.
* A race condition that caused TLS request to wait indefinitely for data to arrive when some was already sitting in the incoming buffer.
